src/Containers/Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -339,68 +343,186 @@ public async Task<string> DownloadBlobAsync(string repository, Descriptor descri
+                hash.AppendData(chunkBackingStore, 0, bytesRead);
+                ByteArrayContent content = new(chunkBackingStore, offset: 0, count: bytesRead);
+                content.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
+                content.Headers.ContentLength = bytesRead;
 
-            HttpResponseMessage patchResponse = await client.PatchAsync(patchUri, content, cancellationToken).ConfigureAwait(false);
+                // manual because ACR throws an error with the .NET type {"Range":"bytes 0-84521/*","Reason":"the Content-Range header format is invalid"}
+                //    content.Headers.Add("Content-Range", $"0-{contents.Length - 1}");
+                Debug.Assert(content.Headers.TryAddWithoutValidation("Content-Range", $"{chunkStart}-{chunkStart + bytesRead - 1}"));
+
+                return new(content, bytesRead);
+            }
+
+            // for a successful chunk, we decrement retries (if any) and read the amount written by the server + next location
+            Action ProcessSuccessfulChunk(HttpResponseMessage response, bool ignoreRange = false) => () =>
+            {
+                if (retryCount > 0)
+                {
+                    retryCount -= 1;
+                }
+                chunkCount += 1;
+                UpdateStateFromResponse(response, ignoreRange);
+            };
 
-            // Fail the upload if the response code is not Accepted (202) or if uploading to Amazon ECR which returns back Created (201).
-            if (!(patchResponse.StatusCode == HttpStatusCode.Accepted || (IsAmazonECRRegistry && patchResponse.StatusCode == HttpStatusCode.Created)))
+            // responses can tell us a) where to upload the next chunk, and b) how much of the last chunk was written.
+            // we use this data to update our internal state before each new iteration.
+            void UpdateStateFromResponse(HttpResponseMessage response, bool ignoreRange)
             {
-                var headers = patchResponse.Headers.ToString();
-                var detail = await patchResponse.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
-                string errorMessage = Resource.FormatString(nameof(Strings.BlobUploadFailed), $"Chunked PATCH {patchUri}", patchResponse.StatusCode, headers + Environment.NewLine + detail);
-                throw new ApplicationException(errorMessage);
+                var amountSent = ignoreRange ? null : ParseRangeAmount(response);
+                patchUri = GetNextLocation(response).Uri;
+                if (amountSent is not null)
+                {
+                    chunkStart = amountSent.Value + 1;
+                    // what the server actually read/processed may not align with the bytesRead from the content originally,
+                    // so seek to that position to ensure no gaps
+                    contents.Seek(chunkStart, SeekOrigin.Begin);
+                }
```

Comment:
This is a pretty big change that's required for correctness - it's possible for registries to only partially accept a chunk, so we need to read their Range headers if present to verify how 'far' in the upload _they_ think they are. If there's a discrepancy, we need to update our read position to that location or else the server will send validation errors.